chain/chain/src/state_sync.rs

```diff
@@ -0,0 +1,191 @@
+    Ok(())
+}
+
+fn on_new_epoch(store_update: &mut StoreUpdate, header: &BlockHeader) -> Result<(), Error> {
+    let num_new_chunks = vec![0usize; header.chunk_mask().len()];
+    store_update.set_ser(DBCol::StateSyncNewChunks, header.hash().as_ref(), &num_new_chunks)?;
+    Ok(())
+}
+
+fn remove_old_epochs(
+    store: &Store,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+    prev_header: &BlockHeader,
+) -> Result<(), Error> {
+    for epoch_id in iter_state_sync_hashes_keys(store) {
+        let epoch_id = epoch_id?;
+        if &epoch_id != header.epoch_id() && &epoch_id != prev_header.epoch_id() {
+            store_update.delete(DBCol::StateSyncHashes, epoch_id.as_ref());
+        }
+    }
+    Ok(())
+}
+
+fn update_sync_hashes<T: ChainStoreAccess>(
+    chain_store: &T,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+) -> Result<(), Error> {
+    let sync_hash = chain_store.get_current_epoch_sync_hash(header.epoch_id())?;
+    if sync_hash.is_some() || header.height() == chain_store.get_genesis_height() {
+        return Ok(());
+    }
+
+    let prev_header = match chain_store.get_block_header(header.prev_hash()) {
+        Ok(h) => h,
+        // During epoch sync, we save headers whose prev headers might not exist, so we just do nothing in this case.
+        // This means that we might not be able to state sync for this epoch, but for now this is not a problem.
+        Err(Error::DBNotFoundErr(_)) => return Ok(()),
+        Err(e) => return Err(e),
+    };
+
+    if prev_header.height() == chain_store.get_genesis_height() {
+        return on_new_epoch(store_update, header);
+    }
+    if prev_header.epoch_id() != header.epoch_id() {
+        on_new_epoch(store_update, header)?;
+        return remove_old_epochs(chain_store.store(), store_update, header, &prev_header);
+    }
+
+    save_epoch_new_chunks(chain_store, store_update, header)?;
+
+    if header.last_final_block() != &CryptoHash::default() {
+        // We don't need to keep info for old blocks around. After a block is finalized, we don't need anything before it
+        let last_final_header = match chain_store.get_block_header(header.last_final_block()) {
+            Ok(h) => h,
+            // This might happen in the case of epoch sync where we save individual headers without having all
+            // headers that belong to the epoch.
+            Err(Error::DBNotFoundErr(_)) => return Ok(()),
+            Err(e) => return Err(e),
+        };
+
+        for block_hash in iter_epoch_new_chunks_keys(chain_store.store()) {
+            let block_hash = block_hash?;
+            let old_header = chain_store.get_block_header(&block_hash)?;
+            if old_header.height() < last_final_header.height() {
+                store_update.delete(DBCol::StateSyncNewChunks, block_hash.as_ref());
+            }
+        }
+    }
+    Ok(())
+}
+
+pub(crate) struct SyncHashTracker<'a> {
```

Comment:
Can you add comment please?

Reply:
actually just deleted this per the comment below